packages/asset-buyer/src/swap_quoter.ts

```diff
@@ -0,0 +1,436 @@
+        const orderProviderRequest = {
+            makerAssetData,
+            takerAssetData,
+            networkId: this.networkId,
+        };
+        const request = orderProviderRequest;
+        // get provider response
+        const response = await this.orderProvider.getOrdersAsync(request);
+        // since the order provider is an injected dependency, validate that it respects the API
+        // ie. it should only return maker/taker assetDatas that are specified
+        orderProviderResponseProcessor.throwIfInvalidResponse(response, request);
+        // process the responses into one object
+        const isMakerAssetZrxToken = makerAssetData === zrxTokenAssetData;
+        const ordersAndFillableAmounts = await orderProviderResponseProcessor.processAsync(
+            response,
+            isMakerAssetZrxToken,
+            this.expiryBufferMs,
+            this._contractWrappers.orderValidator,
+        );
+        const lastRefreshTime = Date.now();
+        const updatedOrdersEntry = {
+            ordersAndFillableAmounts,
+            lastRefreshTime,
+        };
+        this._ordersEntryMap[SwapQuoter._getOrdersEntryMapKey(makerAssetData, takerAssetData)] = updatedOrdersEntry;
+        return ordersAndFillableAmounts;
+    }
+
+    /**
+     * Get the assetData that represents the ZRX token.
+     * Will throw if ZRX does not exist for the current network.
+     */
+    private _getZrxTokenAssetDataOrThrow(): string {
+        return this._contractWrappers.exchange.getZRXAssetData();
+    }
+
+    private async _getSwapQuoteAsync(
+        makerAssetData: string,
+        takerAssetData: string,
+        assetFillAmount: BigNumber,
+        marketOperation: MarketOperation,
+        options: Partial<SwapQuoteRequestOpts>,
+    ): Promise<SwapQuote> {
+        const { shouldForceOrderRefresh, slippagePercentage } = _.merge(
+            {},
+            constants.DEFAULT_SWAP_QUOTE_REQUEST_OPTS,
+            options,
+        );
+        assert.isString('makerAssetData', makerAssetData);
+        assert.isString('takerAssetData', takerAssetData);
+        assert.isBoolean('shouldForceOrderRefresh', shouldForceOrderRefresh);
+        assert.isNumber('slippagePercentage', slippagePercentage);
+        const zrxTokenAssetData = this._getZrxTokenAssetDataOrThrow();
+        const isMakerAssetZrxToken = makerAssetData === zrxTokenAssetData;
+        // get the relevant orders for the makerAsset and fees
+        // if the requested assetData is ZRX, don't get the fee info
+        const [ordersAndFillableAmounts, feeOrdersAndFillableAmounts] = await Promise.all([
+            this.getOrdersAndFillableAmountsAsync(makerAssetData, takerAssetData, shouldForceOrderRefresh),
+            isMakerAssetZrxToken
+                ? Promise.resolve(constants.EMPTY_ORDERS_AND_FILLABLE_AMOUNTS)
+                : this.getOrdersAndFillableAmountsAsync(zrxTokenAssetData, takerAssetData, shouldForceOrderRefresh),
```

Comment:
is this only a requirement for the forwarder?

Reply:
this is part of how fees are taken from the taker if the makerAsset happens to be denominated in the feeToken (in v2's case that would only be the case for ZRX). Later as v3 goes in affect (hopefully), the logic that we conditionally use for `isMakerAssetZrxToken` will probably become the defacto logic for asset-swapper for all asset pairs.